File: AISAM-FE/src/services/postService.ts
import { apiClient } from "@/lib/apiClient";

export type PostStatus = string;

export interface PostItem {
  id: string;
  contentId: string;
  integrationId: string;
  externalPostId: string | null;
  publishedAt: string;
  status: PostStatus;
  contentTitle: string | null;
  brandName: string | null;
  platform: string | null;
  type: string | null;
  caption: string | null;
}

interface GenericResponse<T> {
  success: boolean;
  message?: string | null;
  statusCode?: number;
  data?: T;
  error?: { errorCode?: string; errorMessage?: string };
  timestamp?: string;
}

interface PagedResult<T> {
  data: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

export interface PostFilters {
  page?: number;
  pageSize?: number;
  brandId?: string;
  status?: PostStatus;
}

export async function fetchPosts(params?: PostFilters): Promise<PagedResult<PostItem>> {
  const query = new URLSearchParams();
  if (params?.page) query.set("page", String(params.page));
  if (params?.pageSize) query.set("pageSize", String(params.pageSize));
  if (params?.brandId) query.set("brandId", params.brandId);
  if (params?.status) query.set("status", params.status);

  const res: GenericResponse<PagedResult<PostItem>> = await apiClient(`/posts?${query.toString()}`);
  if (res?.data) return res.data;
  return { data: [], totalCount: 0, page: 1, pageSize: 10, totalPages: 0, hasNextPage: false, hasPreviousPage: false };
}

export async function fetchPost(id: string): Promise<PostItem | null> {
  const res: GenericResponse<PostItem> = await apiClient(`/posts/${id}`);
  if (res?.data) return res.data;
  return null;
}

export async function deletePost(id: string): Promise<boolean> {
  const res: GenericResponse<null> = await apiClient(`/posts/${id}`, { method: "DELETE" });
  return res?.success === true;
}

File: AISAM-BE/AISAM.Services/Service/PostService.cs
using AISAM.Common;
using AISAM.Common.Dtos;
using AISAM.Common.Models;
using AISAM.Data.Enumeration;
using AISAM.Data.Model;
using AISAM.Repositories.IRepositories;
using AISAM.Services.IServices;
using System.Net;

namespace AISAM.Services.Service;

public sealed class PostService : IPostService
{
    private readonly IPostRepository _postRepository;

    public PostService(IPostRepository postRepository)
    {
        _postRepository = postRepository;
    }

    public async Task<GenericResponse<PagedResult<PostListItemDto>>> GetPagedAsync(
        Guid profileId,
        PaginationRequest request,
        Guid? brandId = null,
        ContentStatusEnum? status = null,
        CancellationToken cancellationToken = default)
    {
        var posts = await _postRepository.GetPagedByProfileIdAsync(profileId, request, brandId, status, cancellationToken);

        return GenericResponse<PagedResult<PostListItemDto>>.CreateSuccess(new PagedResult<PostListItemDto>
        {
            Data = posts.Data.Select(MapToDto).ToList(),
            TotalCount = posts.TotalCount,
            Page = posts.Page,
            PageSize = posts.PageSize
        }, "Posts retrieved successfully.");
    }

    public async Task<GenericResponse<PostListItemDto>> GetByIdAsync(Guid profileId, Guid postId, CancellationToken cancellationToken = default)
    {
        var post = await _postRepository.GetByIdAsync(postId, cancellationToken);
        if (post == null || post.IsDeleted || post.Content?.ProfileId != profileId)
        {
            return GenericResponse<PostListItemDto>.CreateError("Post not found.", HttpStatusCode.NotFound);
        }

        return GenericResponse<PostListItemDto>.CreateSuccess(MapToDto(post), "Post retrieved successfully.");
    }

    public async Task<GenericResponse<PagedResult<PostListItemDto>>> GetPagedByWorkspaceAsync(Guid workspaceId, PaginationRequest request, Guid? brandId = null, ContentStatusEnum? status = null, CancellationToken cancellationToken = default)
    {
        var posts = await _postRepository.GetPagedByWorkspaceIdAsync(workspaceId, request, brandId, status, cancellationToken);
        return GenericResponse<PagedResult<PostListItemDto>>.CreateSuccess(new PagedResult<PostListItemDto>
        { Data = posts.Data.Select(MapToDto).ToList(), TotalCount = posts.TotalCount, Page = posts.Page, PageSize = posts.PageSize }, "Posts retrieved successfully.");
    }

    public async Task<GenericResponse<PostListItemDto>> GetByIdInWorkspaceAsync(Guid workspaceId, Guid postId, CancellationToken cancellationToken = default)
    {
        var post = await _postRepository.GetByIdAsync(postId, cancellationToken);
        return post == null || post.IsDeleted || post.Content?.WorkspaceId != workspaceId
            ? GenericResponse<PostListItemDto>.CreateError("Post not found.", HttpStatusCode.NotFound)
            : GenericResponse<PostListItemDto>.CreateSuccess(MapToDto(post), "Post retrieved successfully.");
    }

    private static string MapAdType(AdTypeEnum adType) => adType switch
    {
        AdTypeEnum.TextOnly => "TEXT",
        AdTypeEnum.ImageText => "IMAGE",
        AdTypeEnum.VideoText => "VIDEO",
        _ => adType.ToString()
    };

    private static string? MapPlatform(SocialPlatformEnum? platform) => platform?.ToString().ToLowerInvariant();

    private static PostListItemDto MapToDto(Post post)
    {
        return new PostListItemDto
        {
            Id = post.Id,
            ContentId = post.ContentId,
            IntegrationId = post.IntegrationId,
            ExternalPostId = post.ExternalPostId,
            PublishedAt = post.PublishedAt,
            Status = post.Status.ToString(),
            ContentTitle = post.Content?.Title,
            BrandName = post.Content?.Brand?.Name,
            Platform = MapPlatform(post.Integration?.Platform),
            Type = post.Content != null ? MapAdType(post.Content.AdType) : null,
            Caption = post.Content?.TextContent
        };
    }
}
